src/ott/solvers/linear/univariate.py

```diff
@@ -318,21 +334,52 @@ def quantile_distance(
   y_cdf_inv = all_values_sorted[i_y_cdf_inv]
 
   diff_q = jnp.diff(quantile_levels)
-  cost = jnp.sum(
-      jax.vmap(cost_fn.h)(y_cdf_inv[1:, None] - x_cdf_inv[1:, None]) * diff_q
+  successive_costs = jax.vmap(cost_fn.h)(
+      x_cdf_inv[1:, None] - y_cdf_inv[1:, None]
   )
-  if not return_transport:
-    return cost, None, None
-
-  n = x.shape[0]
-
-  i_in_sorted_x_of_quantile = all_values_sorter[i_x_cdf_inv] % n
-  i_in_sorted_y_of_quantile = all_values_sorter[i_y_cdf_inv] - n
-
-  orig_i = i_x[i_in_sorted_x_of_quantile][1:]
-  orig_j = i_y[i_in_sorted_y_of_quantile][1:]
-
-  return cost, jnp.stack([orig_i, orig_j]), diff_q
+  cost = jnp.sum(successive_costs * diff_q)
+  paired_indices, mass_paired_indices, dual_a, dual_b = [
+      None,
+  ] * 4
+
+  if return_transport or return_dual_vectors:
+    n = x.shape[0]
+
+    i_in_sorted_x_of_quantile = all_values_sorter[i_x_cdf_inv] % n
+    i_in_sorted_y_of_quantile = all_values_sorter[i_y_cdf_inv] - n
+
+    orig_i = i_x[i_in_sorted_x_of_quantile][1:]
+    orig_j = i_y[i_in_sorted_y_of_quantile][1:]
+    paired_indices, mass_paired_indices = jnp.stack([orig_i, orig_j]), diff_q
+
+  if return_dual_vectors:
+    m = y.shape[0]
+    # vector of costs masked by mass transfer. only select active constraints
+    support_cost = (diff_q > 0) * successive_costs
+    # sort and select top n+m-1 to grab non-zero in jit-friendly manner
+    idx_cost = jnp.argsort(support_cost)
+    inv_idx = idx_cost[-n - m + 1:]
+    new_cost = successive_costs[inv_idx]
+    # select indices corresponding to active constraints
+    i_orig_x, i_orig_y = orig_i[inv_idx], orig_j[inv_idx]
+
+    def kkt(dual_ab, ridge_kernel=1e-6):
```

Comment:
local computations (indices `i_orig_x`) appear in the function. in addition, because this is just a sparse linear system, i find it more natural to have it right before the call to the solver